### PR TITLE
[mlir][spirv] Use `AttrTypeReplacer` in map-memref-storage-class. NFC.

### PR DESCRIPTION
**Pull request by kuhar**
Keep the conversion target to allow for checking if the op is legal.

### CODE REVIEW

File: mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp [hunk 1]
```diff
@@ -335,23 +287,25 @@ class MapMemRefStorageClassPass final
...
+    spirv::MemorySpaceToStorageClassConverter converter(spaceToStorage);
+    // Perform the replacement.
+    spirv::convertMemRefTypesAndAttrs(op, converter);
 
-    if (failed(applyFullConversion(op, *target, std::move(patterns))))
+    // Only perform the conversion to check that there are no illegal ops
+    // remaining. Do not attempt to convert anything.
+    if (failed(applyFullConversion(
+            op, *spirv::getMemorySpaceToStorageClassTarget(*context),
+            RewritePatternSet{context})))
```
Comment (qedawkins) [inline mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:308]: I'm not an expert on dialect conversion, but does this just boil down to a walk + ConversionTarget::isLegal? Ditto for the above.

File: mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp [hunk 2]
```diff
@@ -335,23 +287,25 @@ class MapMemRefStorageClassPass final
...
-    spirv::MemorySpaceToStorageClassConverter converter(memorySpaceMap);
-
-    RewritePatternSet patterns(context);
-    spirv::populateMemorySpaceToStorageClassPatterns(converter, patterns);
+    spirv::MemorySpaceToStorageClassConverter converter(spaceToStorage);
+    // Perform the replacement.
+    spirv::convertMemRefTypesAndAttrs(op, converter);
 
-    if (failed(applyFullConversion(op, *target, std::move(patterns))))
+    // Only perform the conversion to check that there are no illegal ops
```
Comment (Groverkss) [inline mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:304]: Can we not do a operation walk instead?

**Review by Groverkss — APPROVED** [review]
LGTM,  just one comment

File: mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp (hunk 1 above)
Comment (kuhar) [reply mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:308]: Yes, exactly. But there's no better named function for this AFAIK. We could simplify this but not exposing a conversion target at all and expect the downstream users to walk this manually.

I'm not sure which one is better honestly -- if we save on some complexity here, we increase complexity downstream.

File: mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp (hunk 2 above)
Comment (kuhar) [reply mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:304]: Same as https://github.com/llvm/llvm-project/pull/80055#discussion_r1471949226

File: mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp (hunk 1 above)
Comment (qedawkins) [reply mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:308]: Can we keep the conversion target and just in these places do
```
op->walk([&](Operation *target) {
  if (!conversionTarget.isLegal(target)) {
    return WalkResult::interrupt();
  }
});
```
?

Comment (antiagainst) [reply mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:308]: Yeah an explicit walk looks better to me--it's simpler and clearer so easy to understand. also we avoid the overhead of conversion framework.

Comment (kuhar) [reply mlir/lib/Conversion/MemRefToSPIRV/MapMemRefStorageClassPass.cpp:308]: OK, I will change this to a walk. But I don't see any complexity in the conversion framework -- it effectively does the same thing.

**Review by antiagainst — APPROVED** [review]

**Review by qedawkins — APPROVED** [review]

**Review by raikonenfnu — APPROVED** [review]
Oh wow, didn't know they introduced a new class for this stuff (AttrTypeReplacer). LGTM :)

**Pull request merged**